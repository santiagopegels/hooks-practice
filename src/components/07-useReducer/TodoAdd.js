import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({handleAddTodo}) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    })
    
    const handleSubmit = (e) => {

        if (description.trim() < 2) {
            return;
        }

        e.preventDefault()

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddTodo(newTodo)
        reset()
    }

    return (
        <>
            <h4>Agregar TODO</h4>
                    <hr></hr>

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Ingrese una tarea"
                            autoComplete="off"
                            value={description}
                            onChange={handleInputChange}
                        />
                        <div className="d-grid gap-2">
                            <button
                                className="btn btn-outline-primary mt-2 btn-block"
                                type="submit"
                            >
                                Agregar
                        </button>
                        </div>
                    </form> 
        </>
    )
}
