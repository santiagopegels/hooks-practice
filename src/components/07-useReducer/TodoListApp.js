import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './todoReducer'

import './style.css'
import { useForm } from '../../hooks/useForm'

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}


export const TodoListApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init)
    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    })


    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])

    const handleSubmit = (e) => {

        if (description.trim() < 2) {
            return;
        }

        e.preventDefault()

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        const action = {
            type: 'add',
            payload: newTodo
        }

        dispatch(action)
        reset()
    }

    return (
        <div>
            <h1>TodoListApp ({todos.length})</h1>
            <hr></hr>

            <div className="row">
                <div className="col-7">

                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, index) => {
                                return <li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p className="text-center">{index + 1}. {todo.desc}</p>
                                    <button className="btn btn-danger">
                                        Delete
                                    </button>
                                </li>
                            })
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar TODO</h4>
                    <hr></hr>

                    <form>
                        <input
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Ingrese una tarea"
                            autoComplete="off"
                            value={description}
                            onChange={handleInputChange}
                        />
                        <div class="d-grid gap-2">
                            <button
                                className="btn btn-outline-primary mt-2 btn-block"
                                type="submit"
                                onClick={handleSubmit}
                            >
                                Agregar
                        </button>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}
