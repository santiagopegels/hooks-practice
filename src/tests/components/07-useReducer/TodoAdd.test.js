import React from 'react'
import { TodoAdd } from '../../../components/07-useReducer/TodoAdd'
import { shallow } from 'enzyme'

describe('Tests on <TodoAdd />', () => {

    const handleAddTodo = jest.fn()

    const wrapper = shallow(
        <TodoAdd
            handleAddTodo={handleAddTodo}
        />
    )

    test('should show correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should not call handleAddTodo', () => {

        const formSubmit = wrapper.find('form').prop('onSubmit')

        formSubmit({ preventDefault() { } });

        expect(handleAddTodo).toHaveBeenCalledTimes(0)
    })

    test('should call handleAddTodo function with an argument', () => {

        const value = 'Learn React'

        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        })

        const formSubmit = wrapper.find('form').prop('onSubmit')

        formSubmit({ preventDefault() { } });

        expect(handleAddTodo).toHaveBeenCalledTimes(1)

        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        })

        //expect that reset function has been executed
        expect( wrapper.find('input').prop('value') ).toBe('')
    })


})
