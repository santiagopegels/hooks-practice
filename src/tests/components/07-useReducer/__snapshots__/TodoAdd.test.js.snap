// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TodoAdd /> should show correctly 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Ingrese una tarea"
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-outline-primary mt-2 btn-block"
        type="submit"
      >
        Agregar
      </button>
    </div>
  </form>
</Fragment>
`;
